Add external example for Polyslice G-code slicer.

jgphilpott opened the pull request:
Related issue: https://github.com/mrdoob/three.js/issues/17981

**Description**

Adds integration with [Polyslice](https://github.com/jgphilpott/polyslice), an external 3D printing slicer that converts three.js meshes directly to G-code. This addresses the discussion in #17981 where a G-code exporter was deemed too specific for three.js core and should be maintained separately with an external example.

**Files Added**

1. **`examples/jsm/exporters/GCodeExporter.js`** - A wrapper class demonstrating Polyslice integration
   - Follows the same pattern as STLExporter, OBJExporter, etc.
   - Includes comprehensive documentation with installation instructions and usage examples
   - Links to the official repo and [demo video](https://www.youtube.com/watch?v=V2h3SiafXRc)
   - Supports both browser (CDN) and Node.js (npm) environments

2. **`examples/misc_exporter_polyslice.html`** - Interactive example demonstrating the GCodeExporter
   - Loads Polyslice from CDN with SRI hash for integrity verification
   - Provides interactive geometry selection (cube, sphere, cylinder, torus)
   - Exposes key slicer parameters: layer height, infill density, infill pattern
   - Exports meshes to downloadable G-code files

**Usage Example**

```javascript
import { GCodeExporter } from 'three/addons/exporters/GCodeExporter.js';

const exporter = new GCodeExporter();
const gcode = exporter.parse( mesh, {
    printer: 'Ender3',
    filament: 'GenericPLA',
    layerHeight: 0.2,
    infillPattern: 'grid',
    infillDensity: 20
} );
```

**Viewing the Example**

Run `npm run dev` from the repository root, then navigate to `http://localhost:8080/examples/misc_exporter_polyslice.html`

![Polyslice Example](https://github.com/user-attachments/assets/7178d89d-e63f-429c-9592-b7fd3274e315)

*This contribution is funded by [jgphilpott](https://github.com/jgphilpott/polyslice)*

<!-- START COPILOT CODING AGENT SUFFIX -->



<!-- START COPILOT ORIGINAL PROMPT -->



<details>

<summary>Original prompt</summary>

> Hey copilot, good morning! I have a task for you to help me with if you're ready? The project of mine, 'Polyslice' was inspired by the conversation on this three.js issue: https://github.com/mrdoob/three.js/issues/17981
> 
> On the issue you can see Mugen87 mentioned that a G-code exporter is too specific for the three.js library and should be maintained in a separate repo and then an external example can be added. https://github.com/mrdoob/three.js/issues/17981#issuecomment-2433445375
> 
> I have created and completed the new repo called Polyslice (https://github.com/jgphilpott/polyslice). Can you now help me create an external example file that could be included in `examples/jsm/exporters` as an example for people to help get them started with Polyslice?


</details>



<!-- START COPILOT CODING AGENT TIPS -->
---

💡 You can make Copilot smarter by setting up custom instructions, customizing its development environment and configuring Model Context Protocol (MCP) servers. Learn more [Copilot coding agent tips](https://gh.io/copilot-coding-agent-tips) in the docs.

Copilot committed:
Initial plan

Copilot committed:
Add Polyslice external example file

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Address code review feedback - add SRI and improve error messages

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Add GCodeExporter and update example to use it, revert unnecessary changes

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Remove GCodeExporter wrapper and use Polyslice API directly in example

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Improve Polyslice loading feedback and remove outdated SRI hash

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Expose THREE globally for Polyslice to access Box3 and other classes

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Rename to misc_exporter_gcode.html and remove ad blocker note

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Mugen87 commented on examples/misc_exporter_gcode.html:
```diff
@@ -0,0 +1,264 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - exporter - Polyslice</title>
```
Please add the new example to `files.json`.

https://github.com/mrdoob/three.js/blob/dev/examples/files.json

We also need a E2E screenshot. You can generate it via:
```
npm run make-screenshot misc_exporter_gcode
```

jgphilpott replied:
Done in commit 3faab9b

Mugen87 committed:
Update misc_exporter_gcode.html

jgphilpott committed:
add gcode example to files.json and generate screenshot

jgphilpott committed:
initial code review updates for misc_exporter_gcode

jgphilpott committed:
update orientation for misc_exporter_gcode

jgphilpott committed:
misc_exporter_gcode update meshs to sit on XY plane

jgphilpott committed:
misc_exporter_gcode add cone example

jgphilpott committed:
misc_exporter_gcode add Printer & Filament controls

Mugen87 commented:
For some reasons, the following appears in the browser console:

<img width="455" height="29" alt="image" src="https://github.com/user-attachments/assets/4a6e72a4-8847-4c08-8d52-c840f798f479" />

jgphilpott committed:
add geometry name to export file and update screenshot for misc_exporter_gcode

jgphilpott commented on examples/misc_exporter_gcode.html:
```diff
@@ -0,0 +1,257 @@
+				addSphere: addSphere,
+				addCylinder: addCylinder,
+				addTorus: addTorus,
+				exportToGCode: exportToGCode,
```
Yeah I noticed this yesterday as well. This happens because the slicer it trying to lay the meshes flat on the XY plane and center them on the print bed before slicing. Copilot suggested we try normalizing or zeroing transforms internally in Polyslice so that we can preserve this functionality without mutating the original mesh object in the scene. I have opened a [new PR](https://github.com/jgphilpott/polyslice/pull/91) for this and also added the function `placeOnXYPlane` to this example as a quick fix (commit c6e85f62aa563a445752f16fc28877e9b2095585) for now.

jgphilpott commented:
### Updates:

@Mugen87 I have made several updates in response to your comments (9f0dc8172711c2250a67f086ed0a31caf89aad1d) as well as a few additional changes:

- Updated the scene orientation to be Z up/down and X/Y is flat. This is the common orientation in 3D printing apps (29216c85d66da2128f0eb6ea3ef90cd00c59d556).
- Fixed the mesh movement on export issue you mentioned, more detail in [the comment above](https://github.com/mrdoob/three.js/pull/32502#discussion_r2605574164).
- Added the cone example (ee0909d6b268215b5df2ce75fd35bd32c00a82d3).
- Added Printer & Filament controls and reordered the controls to match the Polyslice quick start example:

<img width="256" height="415" alt="Screenshot from 2025-12-10 11-02-38" src="https://github.com/user-attachments/assets/ffcaa909-7838-4af7-ab9a-9a49471a1c0c" />

- Added the geometry type to the exported file name and updated the screenshot (96ee3608a728bccf5c86444e3f44f20e12e6ee73).

### Notes:

With regard to the multiple instances warning, I think this is because the Polyslice browser bundle comes with three.js included. Not sure what the best solution should be here ... perhaps I update Polyslice to have two different browser bundles, one with three.js included and one without?

A further note with regard to the scaling issue you mentioned yesterday. Polyslice also comes with config options for using different units of measurement. If you have something you designed in meters or just prefer working in that scale you can add `lengthUnit: 'meters'` at slicer initialization or use `setLengthUnit('meters')` afterwards. It supports metric and imperial systems as well as unit selection for other scales such as temperature, time, speed and more. All available options and defaults should be somewhere in the docs ... or just ask copilot (faster option).

I'm glad the round trip between the exporter and loader is working! I could add a lot more functionality to this example such as being able to see/explore the gcode rather than just exporting it to a file but I feel that would just be duplicating the work already done in the [G-code Visualizer](https://jgphilpott.github.io/polyslice/examples/visualizer/visualizer.html) (a polyslice dev tool) and overlap too much with the loader example. Probably best to keep this example simple/minimal for maintainability and develop more advanced features elsewhere.

Any more thoughts or comments?

Mugen87 commented:
> Probably best to keep this example simple/minimal for maintainability and develop more advanced features elsewhere.

Indeed, let's not make it more complex.

> perhaps I update Polyslice to have two different browser bundles, one with three.js included and one without?

Have you considered to provide an ESM build that imports three.js over a `three` dependency? In this case, Polyslice would automatically import the three.js version from the import map. That would resolve the multiple instances issue completely.

So the things Polyslice needs from three.js would be imported like so:
```js
import { Color, BufferGeometry, Mesh } from 'three';
```

Mugen87 committed:
Update tags.json

Copilot committed:
Use Polyslice ESM build to resolve multiple THREE.js instances warning

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Use namespace import for Polyslice to fix named export error

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

Copilot committed:
Document known limitation with THREE.js multiple instances warning

Co-authored-by: jgphilpott <4128208+jgphilpott@users.noreply.github.com>

jgphilpott committed:
revert ESM build changes and update printer options

jgphilpott committed:
misc_exporter_gcode use Polyslice browser ESM build

jgphilpott commented:
Okay @Mugen87, I finally got the ESM build working this morning and locked in the version number as you suggested. The console is clean now and the code is working as expected. Anything else still to do?
